### PR TITLE
[AArch64][SVE] Improve code quality of vector unsigned/signed add reductions.

### PR DESCRIPTION
**Pull request by dtemirbulatov**
For SVE we don't have to zero extend and sum part of the result before issuing UADDV instruction. Also this change allows to handle bigger than a legal vector type more efficiently and lower a fixed-length vector type to SVE's UADDV where appropriate.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -17455,6 +17455,99 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+      VecOp = convertToScalableVector(DAG, ContainerVT, VecOp);
+    }
+    VecVT = VecOp.getValueType();
+    EVT RdxVT = N->getValueType(0);
+    RdxVT = getPackedSVEVectorVT(RdxVT);
+    SDValue Pg = getPredicateForVector(DAG, DL, VecVT);
+    SDValue Res = DAG.getNode(
+        ISD::INTRINSIC_WO_CHAIN, DL, MVT::i64,
+        DAG.getConstant(Intrinsic::aarch64_sve_uaddv, DL, MVT::i64), Pg, VecOp);
+    EVT ResVT = MVT::i64;
```
[the remark this answers is not in the document]
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17492]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -17455,6 +17455,99 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+  SmallVector<SDValue, 4> Results;
+
+  if (!IsScalableType &&
+      !TLI.useSVEForFixedLengthVectorVT(
+          SplitVals[0].getValueType(),
+          /*OverrideNEON=*/Subtarget.useSVEForFixedLengthVectors(
+              SplitVals[0].getValueType())))
+    return SDValue();
+
+  for (unsigned Num = 0; Num < SplitVals.size(); ++Num) {
```
[the remark this answers is not in the document]
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17530]: Done.

[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17530]: Thanks for combining the two parts. You're still iterating via explicit indexing though, where you could use a range-based for loop for planting both the uaddv instructions and the subsequent scalar adds.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -17455,6 +17455,80 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+      for (SDValue Vec : PrevValues) {
+        SDValue Lo, Hi;
+        std::tie(Lo, Hi) = DAG.SplitVector(Vec, DL);
+        ResultValues.push_back(Lo);
+        ResultValues.push_back(Hi);
+      }
+      if (TLI.isTypeLegal(ResultValues[0].getValueType()))
+        break;
+      PrevValues.clear();
+      std::copy(ResultValues.begin(), ResultValues.end(),
```
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17493]: This seems to be almost reinventing a deque? I think std::deque would be tidier. Get the front item and pop it, perform the split, then push the split results to the back until the front item is legal.

Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17493]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17530]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 4]
```diff
@@ -17455,6 +17456,77 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+      N->getOperand(0).getOpcode() != ISD::SIGN_EXTEND)
+    return SDValue();
+  bool IsSigned = N->getOperand(0).getOpcode() == ISD::SIGN_EXTEND;
+
+  SelectionDAG &DAG = DCI.DAG;
+  auto &Subtarget = DAG.getSubtarget<AArch64Subtarget>();
+  SDValue VecOp = N->getOperand(0).getOperand(0);
+  SDLoc DL(N);
+
+  bool IsScalableType = VecOp.getValueType().isScalableVector();
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17474]: All places where you use `IsScalableType` you invert its value first. So perhaps it makes more sense to have a variable named `IsFixedLengthType = VecOp.getValueType().isFixedLengthVector();`.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 5]
```diff
@@ -17455,6 +17456,77 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+    }
+  }
+
+  EVT ElemType = N->getValueType(0);
+  SmallVector<SDValue, 2> Results;
+  if (!IsScalableType &&
+      !TLI.useSVEForFixedLengthVectorVT(
+          ResultValues[0].getValueType(),
+          /*OverrideNEON=*/Subtarget.useSVEForFixedLengthVectors(
+              ResultValues[0].getValueType())))
```
[the remark this answers is not in the document]
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 6]
```diff
@@ -17455,6 +17456,77 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
...
+  SelectionDAG &DAG = DCI.DAG;
+  auto &Subtarget = DAG.getSubtarget<AArch64Subtarget>();
+  SDValue VecOp = N->getOperand(0).getOperand(0);
+  SDLoc DL(N);
+
+  bool IsScalableType = VecOp.getValueType().isScalableVector();
+  std::deque<SDValue> ResultValues;
+  ResultValues.push_back(VecOp);
+
+  // Split the input vectors if not legal.
```
[the remark this answers is not in the document]
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 4 above)
Comment (dtemirbulatov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done.

**Review by sdesmalen-arm — APPROVED** [review]

**Comment by dtemirbulatov** [comment]
This issue looks like related https://github.com/llvm/llvm-project/issues/98491 to the failure.

**Pull request merged**